[WIP] docs: add example for Angular Universal

steveschmitt opened the pull request:
**WIP - DO NOT MERGE**
This PR should be superseded by #18707 which seemed necessary to overcome merge conflicts.

## PR Checklist
Does please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
```
[ ] Bugfix
[ ] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] angular.io application / infrastructure changes
[ ] Other... Please describe:
```

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
No documentation for Angular Universal

Issue Number: N/A

## What is the new behavior?
Introductory guide and example for Angular Universal

## Does this PR introduce a breaking change?
```
[ ] Yes
[x] No
```

<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

steveschmitt committed:
Move universal example and guide from aio; fix hero service to work on server.

steveschmitt committed:
Move universal example and guide from aio; fix hero service to work on server.

steveschmitt committed:
docs(aio): Universal: fix sample, remove client-side AOT, update doc

steveschmitt committed:
docs(aio): Universal: fix sample, remove client-side AOT, update doc

steveschmitt committed:
docs(aio): Universal: reorganize and fix TOC

steveschmitt committed:
docs(aio): Universal: reorganize and fix TOC

steveschmitt committed:
docs(aio): Universal: remove one set of charts from doc, add images

steveschmitt committed:
docs(aio): Universal: remove one set of charts from doc, add images

steveschmitt committed:
docs(aio): Universal: fix toc and callouts

steveschmitt committed:
docs(aio): Universal: fix toc and callouts

steveschmitt committed:
Merge branch 'aio-universal' of https://github.com/IdeaBlade/angular into aio-universal

steveschmitt committed:
docs(aio): Universal: Extract code examples, add to navigation

mary-poppins commented:
The angular.io preview for 4b868b4440461356b46e6c36792c91ebba080624 is available [here][1].

[1]: https://pr17573-4b868b4440461356b46e6c36792c91ebba080624.ngbuilds.io/

Foxandxss reviewed (changes requested):
I left a few comments, but I feel it needs more work.

wardbell commented:
Let's factor in issues #16038  and #17578

steveschmitt committed:
docs(aio): Universal: Remove startup time measurements; Move webpack config snippets to docregions; Add missing packages; Add package.uni.json

steveschmitt committed:
docs(aio): Add universal example dependencies to boilerplate package.json

steveschmitt commented:
@Foxandxss Have you looked at my changes to this pr?

Foxandxss commented:
Last time I did I still saw a lot of my comments there. Did you look at them? :P

steveschmitt commented on aio/content/guide/universal.md:
```diff
@@ -0,0 +1,496 @@
+
+Angular Universal is a technology that lets Angular applications run outside of the browser.  Using Universal, you can run your Angular application on any [Node.js](https://nodejs.org/en/) or  [ASP.NET Core](https://github.com/MarkPieszak/aspnetcore-angular2-universal) server.  You can use it to generate HTML output from your app on-demand, or generate HTML files ahead of time.  A Universal appli ...
+
+### How does it work? <a name="how-does-it-work"></a>
```
This provides a consistent anchor name for the heading

Foxandxss replied:
We put a `{@a name}` before the header. Please change.

steveschmitt replied:
Done.

Foxandxss replied:
Thank you sir.

steveschmitt committed:
Remove unused graphs

steveschmitt commented on aio/content/guide/universal.md:
```diff
@@ -0,0 +1,496 @@
+To handle requests and return the output, the Universal app is wrapped in a server.  The server (NodeJS or ASP.NET Core) uses the `renderModuleFactory` function from `platform-server` to render the app's components into an internal DOM representation, which is then returned as a string.
+
+The `renderModuleFactory` function uses a *template* HTML page, an Angular *module* containing components, and a *route* that determines which components get rendered.  It renders the components into the `<app>` tag of the template to create the HTML page, similar to the page that would be seen on the client.  The route comes from the HTTP request received by the server.  Each request to the serv ...
+
```
I added a simpler paragraph above, then moved the details into an "implementation" subsection

steveschmitt commented on aio/content/guide/universal.md:
```diff
@@ -0,0 +1,496 @@
+
+Building upon _Tour of Heroes_, the folder structure will look like this:
+
+<code-example format="." language="none" linenums="false">
```
It should work now that I fixed the boilerplate package.json

steveschmitt commented on aio/content/guide/universal.md:
```diff
@@ -0,0 +1,496 @@
+        selector: 'my-component',
+        templateUrl: './my.component.html'
+    })
+</code-example>
```
SystemJS supports full paths, so you might have them in your app if you use SystemJS.  I think the current prose makes that point.

Foxandxss replied:
Reading it again, I don't mind, yes.

steveschmitt commented on aio/content/guide/universal.md:
```diff
@@ -0,0 +1,496 @@
+
+with this:
+
+<code-example path="universal/src/app/app.module.ts" region="browsermodule" title="src/app/app.module.ts (withServerTransition)">
```
Where else does it need to be changed?  In my limited understanding, the name used in `withServerTransition` only affects the stylesheet transition and does not affect the rest of the app.  So it works as long as the same name is used for `withServerTransition` on server and client.

Foxandxss replied:
I don't remember this one right now and there is no preview. I will read everything again and report back.

steveschmitt commented on aio/content/examples/universal/webpack.config.uni.js:
```diff
@@ -0,0 +1,29 @@
+const ngtools = require('@ngtools/webpack');
```
Done.

Foxandxss replied:
docregion, not a docplaster (or both if you really need a docplaster)

Foxandxss replied:
still missing a docregion :P

steveschmitt replied:
Why do I need a docregion at the top of file?  I don't refer to the top section in the doc.  I refer to either the whole file, or to the small docregions that I have defined.

Foxandxss replied:
If you refer to it in the docs (and you do) you need a docregion. A docregion in the top so the tool (dgeni) can see the file and parse it.

wardbell replied:
To be clear, if you want to display the entire file, you should have a nameless `#docregion` at the top. If you only need a fragment, that fragment should have a named `#docregion` and  you do _not_ need a top level `#docregion`.

>That's the way it was and should still be.

steveschmitt commented:
Of course, I looked at them all and made changes where necessary.  I replied (just now) to your comments above.  I also removed the graphs and addressed the issues that Ward requested.

Foxandxss commented:
Good to hear! I didn't peek much because I thought you were still working on it. I will review it shortly :)

steveschmitt committed:
Convert anchors to {@a name} syntax

Foxandxss commented:
Not sure if the preview will work with the conflict. Can you fix it please?

steveschmitt committed:
Merge branch 'master' into aio-universal

steveschmitt committed:
Merge branch 'master' into aio-universal

Foxandxss commented on aio/tools/examples/shared/package.json:
```diff
@@ -32,8 +32,9 @@
-    "@angular/cli": "^1.0.0",
+    "@angular/cli": "^1.2.0",
     "@ngtools/webpack": "^1.2.11",
+    "@angular/cli": "^1.2.0",
```
Why twice?

steveschmitt replied:
Messed up the merge.  Fixed now.

steveschmitt committed:
Update package.json

Fix duplicate @angular/cli

steveschmitt committed:
docs(aio): Universal: remove docplaster from webpack.config.uni.js

steveschmitt committed:
docs(aio): Universal: add docregion to webpack.config.uni.js

steveschmitt committed:
docs(aio): Universal: Fix .gitignore to all webpack.config.uni.js as an example; fix links to other docs in universal.md

mary-poppins commented:
You can preview 4ddeadf at https://pr17573-4ddeadf.ngbuilds.io/.

Foxandxss commented on aio/content/examples/universal/tsconfig-uni.json:
```diff
@@ -15,7 +15,7 @@
 
   "files": [
-    "src/uni/app.server.ts",
+    "src/uni/app.server.module.ts",
```
In this file (not this line) there is a `typeRoots` entry that doesn't need to appear, that is just a need in our docs.

Foxandxss commented on aio/content/examples/universal/webpack.config.uni.js:
```diff
@@ -0,0 +1,37 @@
+const ngtools = require('@ngtools/webpack');
```
remove tabs for 2 spaces please.

Foxandxss commented on aio/content/examples/universal/webpack.config.uni.js:
```diff
@@ -0,0 +1,37 @@
+	target: 'node',
+// #docregion output
+	output: {
+		path: 'src/dist',
```
needs to be inside `src`? can't be outside?

Foxandxss reviewed (changes requested):
The How it works and the implementation part are a bit complicate to follow. The  user comes with no idea and that explanation doesn't help much to see the whole picture. Maybe a diagram?

You need to give a ToH-pt6 zip file. You can use the live-example component for that. That way, the users can follow it step by step.

It also needs e2e tests.

steveschmitt commented on aio/content/guide/universal.md:
```diff
@@ -0,0 +1,507 @@
+
+## Server Code
+
+To run an Angular Universal application on the server, you need a server program that accepts the requests and returns the results.  That's not part of the core Angular app, so you need to add those pieces.
```
How about "the client-side angular app"?

Foxandxss replied:
Sounds good

steveschmitt replied:
Please explain what you mean above, "changing our app name".  What is the app name?

Foxandxss replied:
If I understood it correctly, the `appId` would need to match with something in the client-app, right?

steveschmitt replied:
The AppModule is shared between client and server, so the `appId` declared in `withServerTransition` is used by both.  AFAIK, it doesn't need to be anywhere else.

steveschmitt committed:
docs(aio): Universal: fix missing link and indentation.

mary-poppins commented:
You can preview 698af42 at https://pr17573-698af42.ngbuilds.io/.

m98 commented on aio/content/examples/universal/webpack.config.uni.js:
```diff
@@ -0,0 +1,38 @@
+	target: 'node',
+// #docregion output
+	output: {
+		path: 'src/dist',
```
is not better to be `__dirname + '/src/dist'`?

m98 commented on aio/content/examples/universal/webpack.config.uni.js:
```diff
@@ -0,0 +1,38 @@
+// #docregion rules
+	module: {
+		rules: [
+    		{ test: /\.css$/, loader: 'raw-loader' },
```
What if use is using css processer? (I guess many users are using LESS or SASS)

It would be helpful if we talk a bit about installing sass webpack loader, and after installing it the webpack config should change to:
```

 {
        test: /\.scss$/, use: [{
        loader: "style-loader"
      }, {
        loader: "css-loader"
      }, {
        loader: "sass-loader",
        options: {
          includePaths: ["absolute/path/a", "absolute/path/b"]
        }
      }]
      },
```

m98 commented on aio/content/examples/universal/webpack.config.uni.js:
```diff
@@ -0,0 +1,38 @@
+// #enddocregion plugins
+// #docregion rules
+	module: {
+		rules: [
```
What if use is using css processer? (I guess many users are using LESS or SASS)

It would be helpful if we talk a bit about installing sass webpack loader, and after installing it the webpack config should change to:
```

 {
        test: /\.scss$/, use: [{
        loader: "style-loader"
      }, {
        loader: "css-loader"
      }, {
        loader: "sass-loader",
        options: {
          includePaths: ["absolute/path/a", "absolute/path/b"]
        }
      }]
      },
```

wardbell commented:
Closed by publication of its successor PR #17573
